### ISSUE TITLE
Audio and video are further apart (sync issues)

### ISSUE THREAD

**Issue by antobinary** [bigbluebutton/bigbluebutton#18131]
**Describe the bug**
If you create a recording in BBB 2.6.9 with webcam and microphone, the audio will be playing 2-3 seconds before the video. The gap seems to remain consistent throughout the entire playback

**To Reproduce**
1. Start a meeting with `record=true` on BigBlueButton 2.6.9
2. Join with webcam and microphone.
3. Start recording
4. Count up aloud while counting up with your fingers in front of the camera

**Expected behavior**
By the time you hear "three", your webcam stream shows you holding one finger up.

**Actual behavior**
Smaller, almost negligible delay.

**Examples**
https://test26.bigbluebutton.org/playback/presentation/2.3/b59b2265c47bfe66b81015a37264b27b115dda90-1686068839209
https://test26.bigbluebutton.org/playback/video/b59b2265c47bfe66b81015a37264b27b115dda90-1686068839209/

**Additional context**
BigBlueButton 2.6.9

First reported in https://groups.google.com/g/bigbluebutton-dev/c/VuxFGcPGgrs

**Comment by antobinary** [comment]
I just installed a new server with BBB 2.6**.8** and tried the recordings. There was virtually no lag.
This narrows down the possibilities. Looking at BBB 2.6.9's release notes and checking with @kepstin 
this is likely a side effect from https://github.com/bigbluebutton/bigbluebutton/pull/18044 or https://github.com/bigbluebutton/bbb-webrtc-sfu/compare/v2.9.12...v2.9.13

Also, note that both `presentation` and `video` playbacks display the delay.

**Comment by schrd** [comment]
I did some tests on this one. Video and audio are async if **all** of the following conditions are met:

* 2.6.9 new default recording format is enabled and
* camera is shared before recording starts

A/V is almost in sync if

* 2.6.9 new default recording format is enabled and the camera is shared after recording has started
* `recordFullDurationMedia=true` in `/etc/bigbluebutton/bbb-web.properties`

It is independent from the audio bridge (freeswitch or mediasoup proxy).

**Comment by amg-web** [comment]
@schrd 
what do you **mean 2.6.9 new default recording format is enabled** 
do you mean video downloadable format?
hardly it has matter in our case, it works after meeting

My server does not has this enabled and also affected.

**Comment by antobinary** [comment]
> what do you mean 2.6.9 new default recording format is enabled

I think this was supposed to be "new recording parameter's default value"
`recordFullDurationMedia` from https://github.com/bigbluebutton/bigbluebutton/pull/18044

**Comment by amg-web** [comment]
@antobinary 
my setting is disabled
recordFullDurationMedia=false

**Comment by amg-web** [comment]
@antobinary 

is any idea about nginx errors?:
2023/06/09 10:36:27 [error] 13933#13933: *20507 connect() failed (111: Connection refused) while connecting to upstream, client: 195.64.239.46, server: vclass.com, request: "GET /html5client/locale?locale=en-US&init=true HTTP/2.0", upstream: "http://127.0.0.1:4105/html5client/locale?locale=en-US&init=true", host: "vclass.com", referrer: "https://vclass.com/html5client/join?sessionToken=cywonicyzutpggkj"
2023/06/09 10:36:27 [error] 13933#13933: *20507 connect() failed (111: Connection refused) while connecting to upstream, client: 195.64.239.46, server: vclass.com, request: "GET /html5client/locale?locale=en-US&init=true HTTP/2.0", upstream: "http://127.0.0.1:4106/html5client/locale?locale=en-US&init=true", host: "vclass.com", referrer: "https://vclass.com/html5client/join?sessionToken=cywonicyzutpggkj"

it present evne if only one person is online and just started meeting

**Comment by amg-web** [comment]
users also complain about sessions without recording

**Comment by kepstin** [comment]
@amg-web The nginx errors and audio-video sync issues in the live session are unrelated to this issue.

Due to the way media works in BigBlueButton - with audio and video streamed separately - there's no good way to ensure that they're perfectly in sync in the live session, and if there is a desync in the live session, that will probably also be seen in the recording. This issue is about a desync that happens only in the recording.

**Comment by amg-web** [comment]
@kepstin 
yes agree. sorry.

does anyone who have node exporter 0.6.1 experience av sync issue?

**Comment by Daniel-at-git** [comment]
> does anyone who have node exporter 0.6.1 experience av sync issue?

Using node exporter 0.6.1 and we don't have av sync issues in 2.6.9. 
We start recordings after camera has shared....

**Comment by amguirado73** [comment]
Hello,

I just ran a test on test and it works fine. Has any change been made?

https://test26.bigbluebutton.org/playback/presentation/2.3/3e2d01ed8c5d98b144ce3ae8cf22551554240965-1687175833730

Thanks

**Comment by ffdixon** [comment]
Digging into this deeper.  We just did a test on a newly installed 2.6.8, 2.6.9, and 2.6.10 server, and confirmed the audio drift was present in 2.6.9 but **not** in 2.6.10.

https://github.com/bigbluebutton/bigbluebutton/assets/290351/ee590440-942b-4861-9a3e-d97baa25895a

I shared audio and video before starting recording, which is what would occur in a typical session.

The underlying hardware on the servers is the same.  I also confirmed that `recordFullDurationMedia=false` is set on the 2.6.9 and 2.6.10.

Is there anyone else that is running 2.6.9 (and experienced the drift) and upgraded to 2.6.10 and found new recordings are back in sync.

**Comment by amguirado73** [comment]
Hello, I can still reproduce the issue in 2.6.10.

**Comment by kepstin** [comment]
There were no changes made in BigBlueButton 2.6.10 that were relevant to this issue. Because of the inconsistency of the issue, due to the details of what causes it, a single working recording is not enough to say that it has been fixed.

**Comment by ffdixon** [comment]
@amguirado73 Can you give us steps to reproduce, specifically the sequence of sharing audio, video, and starting sharing.

**Comment by ffdixon** [comment]
Digging into this more.  The parameter `recordFullDurationMedia=false` is false by default, which means it **does not** record all the media for the session and then process it afterwards (if there was a recording).  

  https://github.com/bigbluebutton/bigbluebutton/blob/v2.6.10/bigbluebutton-web/grails-app/conf/bigbluebutton.properties#L237

The default behaviour switched to only recording media when the Start/Stop Record button is pressed.  This is more desirable, but has the hit-and-miss synchronization issue.  

So, setting`recordFullDurationMedia=true` will make recording (and timestamps) work as 2.6.8, and should give you the synchronization as in 2.6.8.  

We're digging into why having `recordFullDurationMedia=false` is causing sync issues in some cases.

**Comment by amguirado73** [comment]
Hello,

I just performed 3 tests and it seems that the issue no longer happens.

Thanks 
Regards

**Comment by amguirado73** [comment]
Yes, i have applied the patch and everything seems to be fine.

Regards

**Comment by jarrodjay** [comment]
In my short testing, I feel like the patch definitely helped, with 2-3 second gap in last week's recordings, changing to 0.5s in today's recordings.
Looking at the patch itself, that's not going to help with rebuilding past recordings, correct?

**Comment by Stoppschildgruen** [comment]
The problem is still occurring. Recordings are often out of sync even with v2.6.11 and also with v2.7 RC1. With v2.5.18 it is in sync.

**Comment by JeanPluzo** [comment]
@Stoppschildgruen, have you installed and tested v2.6.12? It came out just yesterday.

**Comment by Stoppschildgruen** [comment]
Yes, same with v2.6.12 - recordings are still mostly out of sync. Of three recordings, only the first was in sync. We would like to update our productive system from v2.5 to v2.6, but unfortunately this is holding us back.

**Comment by defnull** [comment]
I just tested this with 2.6.12 and have a 1-2 second delay even with `recordFullDurationMedia=true` in `/etc/bigbluebutton/bbb-web.properties` in both presentation and video formats :/

**Comment by Stoppschildgruen** [comment]
With `recordFullDurationMedia=true` it works and all recordings made for test are in sync. Thanks!

**Comment by Ithanil** [comment]
> I just tested this with 2.6.12 and have a 1-2 second delay with `recordFullDurationMedia=true` in `/etc/bigbluebutton/bbb-web.properties` in both presentation and video formats :/ Testing again with `recordFullDurationMedia=false` removes the delay.
> 
> I assume the patch from above fixed the new `recordFullDurationMedia=false` behavior (in most but not all cases), but broke the old `recordFullDurationMedia=true` behavior in return. Is that possible?

@prlanzarin FYI: We still receive reports about desync recordings with `recordFullDurationMedia=false`, albeit the issue seems to appear less often. And I was quite happy to be able to drop some hacky custom post-processing to simulate what  `recordFullDurationMedia=false` does. Simply setting it to true is not really an option.

Just wanted to let you know that this is not finally solved.
